File: php/src/CodingInterview/Application/Usecase/Order/AdditionalBuyOrderUsecase.php
<?php

declare(strict_types=1);

namespace Folio\CodingInterview\Application\Usecase\Order;

use Folio\CodingInterview\Application\Repository\AccountRepository;
use Folio\CodingInterview\Application\Repository\MarketPriceRepository;
use Folio\CodingInterview\Application\Repository\PortfolioRepository;
use Folio\CodingInterview\Application\Service\PortfolioService;
use Folio\CodingInterview\Domain\AppConstants;
use Folio\CodingInterview\Domain\BigDecimal;
use Folio\CodingInterview\Domain\UserId;

final class AdditionalBuyOrderUsecaseInput
{
    public function __construct(
        public readonly UserId $userId,
        public readonly BigDecimal $amount,
    ) {}
}

final class AdditionalBuyOrderUserNotFoundException extends \RuntimeException
{
    public function __construct() { parent::__construct('user has no live account'); }
}

final class AdditionalBuyOrderAmountTooSmallException extends \RuntimeException
{
    public function __construct() { parent::__construct('amount is too small'); }
}

final class AdditionalBuyOrderUsecase
{
    public function __construct(
        private readonly AccountRepository $accountRepository,
        private readonly PortfolioRepository $portfolioRepository,
        private readonly MarketPriceRepository $marketPriceRepository,
    ) {}

    public function run(AdditionalBuyOrderUsecaseInput $input): void
    {
        if ($input->amount->lt(AppConstants::minOperationAmount())) {
            throw new AdditionalBuyOrderAmountTooSmallException();
        }
        $account = $this->accountRepository->find($input->userId);
        if ($account === null) {
            throw new AdditionalBuyOrderUserNotFoundException();
        }
        $portfolio = $this->portfolioRepository->get();
        $prices = $this->marketPriceRepository->all();
        $updated = PortfolioService::allocateAdditional($account, $input->amount, $portfolio, $prices);
        $this->accountRepository->upsert($input->userId, $updated);
    }
}

File: java8/src/main/java/folio/codinginterview/application/usecase/order/AdditionalBuyOrderUsecase.java
package folio.codinginterview.application.usecase.order;

import folio.codinginterview.application.repository.AccountRepository;
import folio.codinginterview.application.repository.MarketPriceRepository;
import folio.codinginterview.application.repository.PortfolioRepository;
import folio.codinginterview.application.service.PortfolioService;
import folio.codinginterview.domain.Account;
import folio.codinginterview.domain.AppConstants;
import folio.codinginterview.domain.UserId;

import java.math.BigDecimal;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;

public final class AdditionalBuyOrderUsecase {
    public static final class Input {
        private final UserId userId;
        private final BigDecimal amount;
        public Input(UserId userId, BigDecimal amount) {
            this.userId = userId;
            this.amount = amount;
        }
        public UserId userId() { return userId; }
        public BigDecimal amount() { return amount; }
    }

    public static class Exception extends RuntimeException {
        protected Exception(String msg) { super(msg); }
    }

    public static final class UserNotFound extends Exception {
        public static final UserNotFound INSTANCE = new UserNotFound();
        private UserNotFound() { super("user not found"); }
    }

    public static final class AmountTooSmall extends Exception {
        public static final AmountTooSmall INSTANCE = new AmountTooSmall();
        private AmountTooSmall() { super("amount too small"); }
    }

    private final AccountRepository accountRepository;
    private final PortfolioRepository portfolioRepository;
    private final MarketPriceRepository marketPriceRepository;

    public AdditionalBuyOrderUsecase(
            AccountRepository accountRepository,
            PortfolioRepository portfolioRepository,
            MarketPriceRepository marketPriceRepository
    ) {
        this.accountRepository = accountRepository;
        this.portfolioRepository = portfolioRepository;
        this.marketPriceRepository = marketPriceRepository;
    }

    public CompletableFuture<Void> run(Input input) {
        if (input.amount().compareTo(AppConstants.MIN_OPERATION_AMOUNT) < 0) {
            CompletableFuture<Void> failed = new CompletableFuture<>();
            failed.completeExceptionally(AmountTooSmall.INSTANCE);
            return failed;
        }
        return accountRepository.find(input.userId()).thenCompose((Optional<Account> maybeAccount) -> {
            if (!maybeAccount.isPresent()) {
                CompletableFuture<Void> failed = new CompletableFuture<>();
                failed.completeExceptionally(UserNotFound.INSTANCE);
                return failed;
            }
            Account account = maybeAccount.get();
            return portfolioRepository.get().thenCompose(portfolio ->
                    marketPriceRepository.all().thenCompose(prices -> {
                        Account updated = PortfolioService.allocateAdditional(account, input.amount(), portfolio, prices);
                        return accountRepository.upsert(input.userId(), updated);
                    }));
        });
    }
}
